UnityProject/Assets/Scripts/Actions/V2/ActionManager.cs

```diff
@@ -0,0 +1,321 @@
+		{
+			// Notify UI system on client to refresh buttons
+			UpdateMe();
+		}
+
+		public void RegisterNewAction(ActionButtonData newData, Action<Vector2> logic)
+		{
+			newData.TrackingObject = gameObject.NetWorkIdentity();
+			switch (newData.TriggerType)
+			{
+				case ActionTriggerType.ServerOnly:
+					ServerAddAction(newData, logic);
+					break;
+				case ActionTriggerType.ClientOnly:
+					ClientAddAction(newData, logic);
+					break;
+				case ActionTriggerType.Both:
+				default:
+					ServerAddAction(newData, logic);
+					ClientAddAction(newData, logic);
+					break;
+			}
+			ActionButtons.Add(newData);
+		}
+
+		/// <summary>
+		/// Registers a new action with the given parameters.
+		/// </summary>
+		/// <param name="newID">the ID that will be used to trigger said command</param>
+		/// <param name="displayName">The UI name</param>
+		/// <param name="desc">The description of the action that will be executed</param>
+		/// <param name="triggerType">Do you want to execute this command on the server? the client? or both? (Server always recommended)</param>
+		/// <param name="Icon">The graphics used for the button</param>
+		/// <param name="logic">The actual function that will be run when pressing the button</param>
+		/// <param name="canBeUsedWhileGhosting">[Mind Action Manger Only] - Can this action be used while ghosting?</param>
+		/// <param name="cooldownTime">How long before we can run this command again?</param>
+		public void RegisterNewAction(string newID, string displayName, string desc, ActionTriggerType triggerType,
+			List<SpriteDataSO> Icon, Action<Vector2> logic, bool canBeUsedWhileGhosting = false, float cooldownTime = 0f)
+		{
+			var ActionData = new ActionButtonData
+			{
+				ID = newID,
+				DisplayName = displayName,
+				Description = desc,
+				TriggerType = triggerType,
+				CooldownTime = cooldownTime,
+				AnimatedIconCatalogue = Icon,
+				CanUseWhileGhosting = canBeUsedWhileGhosting,
+				TrackingObject = gameObject.NetWorkIdentity()
+			};
+
+			switch (triggerType)
+			{
+				case ActionTriggerType.ServerOnly:
+					ServerAddAction(ActionData, logic);
+					break;
+				case ActionTriggerType.ClientOnly:
+					ClientAddAction(ActionData, logic);
+					break;
+				case ActionTriggerType.Both:
+				default:
+					ServerAddAction(ActionData, logic);
+					ClientAddAction(ActionData, logic);
+					break;
+			}
+			ActionButtons.Add(ActionData);
+		}
+
+		public void UnregisterAction(ActionButtonData data)
+		{
+			switch (data.TriggerType)
+			{
+				case ActionTriggerType.ServerOnly:
+					ServerRemoveAction(data.ID);
+					break;
+				case ActionTriggerType.ClientOnly:
+					ClientRemoveAction(data.ID);
+					break;
+				case ActionTriggerType.Both:
+				default:
+					ServerRemoveAction(data.ID);
+					ClientRemoveAction(data.ID);
+					break;
+			}
+		}
+
+		[Command]
+		public void CmdTriggerAction(string actionId, Vector2 mouseLocation)
```

Comment:
Should be probably a  net message or try catch

Reply:
Ghost and Player_V4 owns them

also if you want net messages, use the `ClientOnly` trigger and handle them yourself.

Reply:
cool cool